requests-for-comments/2022-06-01-session-analysis.md

```diff
@@ -0,0 +1,118 @@
+2. Building out a new sessions feature tailor made for this analysis
+
+(tl;dr we should do option 1 because we can lean on lots of existing behavior of trends)
+
+The big trade-offs between these two options are:
+
+- Option 1 allows for lots of analysis that you “get for free” because of the power of trends. For example the following are all easily possible:
+    - Plot the number of sessions over time along with the count of users. Then use formulas to divide the 2 and get “sessions per user”
+    - Plot average session length where a users does Z broken down by browser, and notice that your feature is broken on IE.
+    - Plot the average session length where users do X and compare it to the average session length where users do Y. Notice that users that do X, stick around longer.
+    - Run an experiment where the goal is to increase the session length 🤯
+- Potentially, option 2 would allow for more complex sessions to be defined. Right now, the language of trends is about defining events, not sessions. As a result, when this language is translated to sessions, it means, “Give me sessions where this specific event occurs”. This restricts users from creating complex session definitions such as “sessions where X happens and the Y happens within 2 minutes”.
+
+My vote is that we **do option 1 and build this analysis into the existing trends product**. The benefits of building these features into trends greatly outweighs the complex sessions a user could define in option 2. Also, there are future paths where we could allow users to define complex sessions within option 1.
+
+## How should these added analyses be built into trends?
+
+Looking at the various analyses that we want to allow, it seems like the requirements above could fit relatively cleanly into our existing insights UX. Below is a proposal for how we would enable all of the above defined requirements.
+
+### 1. Number of sessions where specific events occur over time
+
+Proposal: Inside trends, we add “Unique Session” as an ‘event aggregation’ option.
+
+![image](https://user-images.githubusercontent.com/4813045/171521247-45dfc5d4-d2c5-4ca4-9fec-f81cc0ace5e8.png)
+
+
+### 2. Average session length over time
+
+Proposal: We allow users to do property aggregations based on “session length”
+
+![image](https://user-images.githubusercontent.com/4813045/171521195-121300f6-48a9-488c-956c-f176243865a7.png)
+
+Details:
+- In addition to showing event properties in the taxonomic filter when a property aggregation is selected, we also show “session properties”. Within session properties, we would show session length.
+- If a session property is the target of the property aggregation, then the query would be aggregated by sessions.
+
+### 3. Average number of sessions per user over time
+
+Proposal: Users accomplish this using formulas. They would make a series that is (a) Session count and (b) user count, and then divide (a) by (b).
```

Comment:
Some driveby feedback from a user here. I feel like 'sessions per user' is too important a metric to be reying on formulas. 

It makes sense logically, but I can easily imagine some users will need help disovering the solution to how you get this metric, when it ought to be obvious and require no cognitive load on the user.

Reply:
Fair point. More generally, 'users per organisation', 'paymentIDs per user', 'paymentIDs per organisation', i.e. count / math operation on one group type, aggregated by another group type would be a really nice to have.

I think it's possible, but raises a few code complications (which would get less complicated once person-on-events exists), so I agree with the decision to push it down to the future, specially since there's a "hack" available to get the same results, and there's an open question of how the UI for this should work.